scale/recipe/models.py

```diff
@@ -324,6 +333,64 @@ def get_details(self, recipe_id):
+        """Schedules an existing recipe for re-processing. All requested jobs, jobs that have changed in the latest
+        revision, and any of their dependent jobs will be re-processed. All database changes occur in an atomic
+        transaction.
+
+        :param recipe_id: The identifier of the recipe to re-process.
+        :type recipe_id: int
+        :param job_names: A list of job names from the recipe that should be forced to re-process even if the latest
+            recipe revision left them unchanged. If none are passed, then only jobs that changed are scheduled.
+        :type job_names: [string]
+        :param all_jobs: Indicates all jobs should be forced to re-process even if the latest recipe revision left them
+            unchanged. This is a convenience for passing all the individual names in the job_names parameter and this
+            parameter will override any values passed there.
+        :type all_jobs: bool
+        :returns: A handler for the new recipe
+        :rtype: :class:`recipe.handlers.handler.RecipeHandler`
+        """
+
+        # Determine the old recipe graph
+        prev_recipe = Recipe.objects.select_related('recipe_type', 'recipe_type_rev').get(pk=recipe_id)
+        prev_graph = prev_recipe.get_recipe_definition().get_graph()
+
+        # Populate the list of all job names in the recipe as a shortcut
+        if all_jobs:
+            job_names = prev_graph.get_topological_order()
+
+        # Determine the current recipe graph
+        current_type = prev_recipe.recipe_type
+        current_graph = current_type.get_recipe_definition().get_graph()
+
+        # Make sure that something is different to reprocess
+        if current_type.revision_num == prev_recipe.recipe_type_rev.revision_num and not job_names:
+            raise ReprocessError('Superseded jobs must be specified when the recipe type has not changed')
+
+        # Compute the job differences between recipe revisions including forced ones
+        graph_delta = RecipeGraphDelta(prev_graph, current_graph)
+        if job_names:
+            for job_name in job_names:
+                graph_delta.reprocess_identical_node(job_name)
+
+        # Get the old recipe jobs that will be superseded
+        prev_recipe_jobs = RecipeJob.objects.filter(recipe=prev_recipe)
+
+        # Acquire model locks
+        superseded_recipe = Recipe.objects.select_for_update().get(pk=recipe_id)
+        prev_jobs = Job.objects.select_for_update().filter(pk__in=[rj.job_id for rj in prev_recipe_jobs])
+        prev_jobs_dict = {j.id: j for j in prev_jobs}
+        superseded_jobs = {rj.job_name: prev_jobs_dict[rj.job_id] for rj in prev_recipe_jobs}
+
+        # Create an event to represent this request
+        description = {'recipe_id': recipe_id, 'job_names': job_names}
+        event = TriggerEvent.objects.create_trigger_event('REPROCESS_RECIPE', None, description, timezone.now())
```

Comment:
I think we should use the same "user" trigger event as we do for the REST APIs for queuing new jobs and recipes. The trigger event is not so much about what we are doing (re-processing), but about what caused this to happen (user/API call, a Scale Strike process, etc.) All user-initiated events (API calls) should be grouped together.